Add new omrsysinfo_get_CPU_load API

fjeremic opened the pull request:
This PR implements the API proposed in option 5. in [1]. The new API is implemented on all platforms currently supported by OMR. The new `omrsysinfo_get_CPU_load` calculates the cumulative CPU utilization of all CPUs on the system as a double in the [0.0, 1.0] interval. We implement the new API using the already existing API `omrsysinfo_get_CPU_utilization` by taking two snapshots of CPU time
usage over time and calculating the delta.

At the moment the new `omrsysinfo_get_CPU_load` API is not thread safe.

[1] https://github.com/eclipse/omr/pull/3152#issuecomment-438347430

fjeremic commented:
Suggesting @charliegracie and @dnakamura for review since they were originally involved in the design discussions in #3152. I'm also subscribing @smlambert / @llxia on this on as I have some test questions:

1. There seems to be duplication of port library tests (`sysinfo_test_get_CPU_utilization`) between OMR and OpenJ9
    - https://github.com/eclipse/omr/blob/2b78ad9e88dcf03cf5c1d0734f996aabcf2bc985/fvtest/porttest/si.cpp#L1592-L1640
    - https://github.com/eclipse/openj9/blob/bd7c4022a8d6efafe277853c28b4528f470a64f7/runtime/tests/port/si.c#L1396-L1443
    - Ideally we should eliminate this duplication, so should all the tests sink down into OMR?
2. There seems to be additional (superset?) of testing of CPU load in `omrtiGetSystemCpuLoad`
    - https://github.com/eclipse/omr/blob/2b78ad9e88dcf03cf5c1d0734f996aabcf2bc985/omr/OMR_TI.cpp#L403-L477
    - I could not find any uses of this code though. Does anyone know how it is invoked and whether this test should be extended?
3. What is the best way to implement tests for this new port library API? Do we have examples of this from past experiences I can mimic?

fjeremic commented:
@charliegracie @dnakamura just making sure this ends up on your radar as you were involved in earlier discussions around this issue. Also subscribing @youngar and @rwy0717 as well since they too are familiar with the port library.

rwy0717 commented:
IIUC, the first call to this API will always fail (on unix/windows only). Can we initialize the latest/oldest cpu time on startup, rather than during the first call?

rwy0717 commented:
I'm going to assign this to myself, but if anyone else wants to take over, feel free to reassign.

fjeremic commented:
> IIUC, the first call to this API will always fail (on unix/windows only). Can we initialize the latest/oldest cpu time on startup, rather than during the first call?

We can do this, however there is a tradeoff. If we initialize the oldest CPU load at startup we run the risk of a user calling the API just once at some much later point. In this case the denominator in the calculation in [1] will be huge, since a lot of time has passed between startup and the user calling the API, and as such the reported CPU load will likely be very close to zero.

Is this something we are willing to accept? I'm indifferent to either option.

[1] https://github.com/eclipse/omr/pull/4702/files#diff-1190c3bb4f08d524f88641310cfd94e0R3560

rwy0717 commented on port/common/omrsysinfo.c:
```diff
@@ -585,6 +585,24 @@ omrsysinfo_get_CPU_utilization(struct OMRPortLibrary *portLibrary, struct J9Sysi
+ * 0.0 means all of the CPUs were idle in the recent period of time observed, while a value of 1.0 means that all CPUs
+ * were actively utilized 100% of the time during the recent period of time observed.
+ *
+ * All values in the inverval are possible.
```
```suggestion
 * All values in the interval are possible.
```

fjeremic replied:
Fixed in 3624cc9

rwy0717 commented:
re 3: You can add the test to `fvtest/porttest/si.cpp`. There are other tests in that file you might use as a model, but I have some advice:

* Use gtest in a normal way (`ASSERT_EQ`, `EXPECT_TRUE`, etc). Do not use the custom loggers or error reporting (`outputErrorMessage`, `reportTestExit`, etc). These custom loggers were ported from J9, and tacked onto gtest, but they're not needed.

* Writing tests for APIs like this are difficult, because you cannot reliably validate the results, you can only exercise the code. Historically we use work and sleep functions to try and put results in an anticipated range. IMHO this makes the test flakey and slow.

fjeremic commented:
@genie-omr build all

fjeremic commented:
@genie-omr build all

fjeremic commented:
@genie-omr build arm

fjeremic commented:
@genie-omr build all

fjeremic commented:
@rwy0717 I've updated the documentation and added a unit test. This is ready for another review.

Before:
```
16:21:22  14: [----------] 38 tests from PortSysinfoTest
16:21:22  14: originalSoftLimit=1048576
16:21:22  14: originalHardLimit=1048576
16:21:22  14: soft set to hard limit=1048576
16:21:29  14: [----------] 38 tests from PortSysinfoTest (6928 ms total)
```

After:
```
12:42:35  14: [----------] 39 tests from PortSysinfoTest
12:42:35  14: originalSoftLimit=1048576
12:42:35  14: originalHardLimit=1048576
12:42:35  14: soft set to hard limit=1048576
12:42:43  14: [----------] 39 tests from PortSysinfoTest (7404 ms total)
```

fjeremic commented:
@rwy0717 are the new tests sufficient?

rwy0717 commented:
Looks good to me!

fjeremic commented:
Thanks! We should give @dnakamura and @charliegracie a chance to review if they wish.

fjeremic commented:
@rwy0717 unless you want to give more time I think this one should be ready.

rwy0717 commented:
Agreed.
